### DISCUSSION TITLE
Insufficient write permission to the media root

### DISCUSSION THREAD

**Discussion by dejantep** [netbox-community/netbox#9567] [Q&A] [answered]
Hello everybody!

I'm having issues with image upload and i'm stuck. I think my permissions are correct but obviously not. Been looking for other threads for clues but no luck.  This is what i get:

```
<class 'PermissionError'>

[Errno 13] Permission denied: '/opt/netbox/netbox/media/devicetype-images/huawei-ups2000g.PNG'

Python version: 3.8.10
NetBox version: 3.2.3
```

When i check folder and files everything seems fine or at least exactly as my other netbox installation. Here's a snippet:

```
dejan@netbox3:~$ ls -l /opt/netbox/netbox/media/
total 16
drwxr-xr-x 2 www-data www-data  4096 Nov 18  2021 devicetype-images
drwxr-xr-x 2 www-data www-data 12288 May  4 08:29 image-attachments
dejan@netbox3:~$
```

If i look at the devicetype-images folder it looks like this:

```
-rw-r--r-- 1 www-data www-data  35306 Feb  1  2021 WS-C2940-8TT-S.jpg
-rw-r--r-- 1 www-data www-data  30138 Feb  1  2021 WS-C2960G-24TC-L.jpg
-rw-r--r-- 1 www-data www-data  37555 Feb  1  2021 WS-C2960G-8TC-L.jpg
-rw-r--r-- 1 www-data www-data  32142 Feb  1  2021 WS-C2960X-24TS-L.jpg
```

What do i need to do?

**Comment by jeremystretch** [comment, marked as answer]
The directory must be writable by the user as which NetBox runs (typically named `netbox`). You can try assigning this user to the `www-data` group and changing the directory to chmod 775 to grant write access.

**Reply by dejantep** [reply]
Thank you Jeremy. It works after suggested changes

**Reply by darrenwoods3** [reply]
How do I determine the user that is being used by netbox?  I went through the installation guide and thought I used the username netbox but that user is not in the /etc/passwd file.

**Reply by candlerb** [reply]
How are you starting Netbox?  If it's via systemd as per the current installation instructions, then the user will be set in the [service file](https://github.com/netbox-community/netbox/blob/master/contrib/netbox.service#L10-L11).

If this Netbox was installed a long time ago - by which I mean more than 2-3 years ago - then older instructions showed running it as "www-data" user, since this already exists on many systems.  And in that case you may also find Netbox being started from supervisord rather than systemd.

If your Netbox is actually working, you can always just look at running processes: the first column will give the user ID.

```
ps auxwww | grep netbox
```

You're not using docker are you?  If so, obviously no user will be created on your host.  Inside the container, I'm not sure - it looks like it might be using a user ["unit"](https://github.com/netbox-community/netbox-docker/blob/release/docker/launch-netbox.sh#L55)

**Reply by JMB1892** [reply]
Hi where would you find the www-data group to set the permissions?

**Reply by candlerb** [reply]
"www-data" is a user and group on your system, typically created by the apache2 package in a Debian/Ubuntu system. You'll find it listed in /etc/passwd and /etc/groups.

Old installation instructions for Netbox told you to install supervisord and to run gunicorn as this "www-data" user.

Current installation instructions recommend that you create a new user "netbox" to run the daemon as. You'll find that as one of the steps [here](https://docs.netbox.dev/en/stable/installation/3-netbox/#create-the-netbox-system-user). They also show running it under systemd.

You haven't described what your actual problem is though. If you have problems with Unix perms and are unclear how these work, I think the best course of action would be to find a local Unix/Linux admin who can help you.

**Reply by JMB1892** [reply]
Thanks so the issue I have is as per the original thread where we attempt to add an image to a rack and get 'insufficient write permissions to the media root' error message.
The permissions on the folder /opt/netbox/netbox/media/image-attachments/ looks correct so it was suggested the netbox user needs to be given additional permissions.

Regards

**Reply by candlerb** [reply]
> The permissions on the folder /opt/netbox/netbox/media/image-attachments/ looks correct

Can you demonstrate that?

```
ls -ld /opt/netbox/netbox/media/image-attachments
```

And compare to the userid that netbox is running as:

`ps auxwww | grep gunicorn`

The parent directories also need to be accessible (at least 'x' perms to world or the netbox user)

```
ls -ld /opt/netbox/netbox/media
ls -ld /opt/netbox/netbox
ls -ld /opt/netbox
ls -ld /opt
```

**Comment by darrenwoods3** [comment]
Brian,

Thanks for your response. This is a brand new install.  The service file does show that it’s using the user netbox.  Any help would be greatly appreciated.

[Unit]
Description=NetBox WSGI Service
Documentation=https://docs.netbox.dev/
After=network-online.target
Wants=network-online.target

[Service]
Type=simple

User=netbox
Group=netbox
PIDFile=/var/tmp/netbox.pid
WorkingDirectory=/opt/netbox

ExecStart=/opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi

Restart=on-failure
RestartSec=30
PrivateTmp=true

[Install]
WantedBy=multi-user.target

***@***.***:/opt/netbox/contrib$ ps auxwww | grep netbox
www-data    1021  0.0  0.1  29632 21988 ?        S    01:17   0:07 /usr/bin/python3 /usr/local/bin/gunicorn -c /opt/netbox/gunicorn_config.py netbox.wsgi
www-data    1072  0.0  1.0 208176 170384 ?       S    01:17   0:08 /usr/bin/python3 /usr/local/bin/gunicorn -c /opt/netbox/gunicorn_config.py netbox.wsgi
www-data    1105  0.0  1.1 222600 184708 ?       S    01:17   0:26 /usr/bin/python3 /usr/local/bin/gunicorn -c /opt/netbox/gunicorn_config.py netbox.wsgi
www-data    1106  0.0  1.0 204764 172696 ?       S    01:17   0:15 /usr/bin/python3 /usr/local/bin/gunicorn -c /opt/netbox/gunicorn_config.py netbox.wsgi
postgres   51161  0.0  0.1 218160 23788 ?        Ss   16:05   0:00 postgres: 12/main: netbox netbox 127.0.0.1(60018) idle
postgres   51181  0.0  0.1 219480 24568 ?        Ss   16:06   0:00 postgres: 12/main: netbox netbox 127.0.0.1(60022) idle
postgres   51334  0.0  0.1 218020 24848 ?        Ss   16:09   0:00 postgres: 12/main: netbox netbox 127.0.0.1(60046) idle
root       70990  123  0.8 180428 139184 ?       R    21:49   0:02 python3 /opt/netbox/netbox/manage.py rqworker directory = /opt/netbox/netbox/ user = www-data
dwoods3+   70994  0.0  0.0   6432   720 pts/0    S+   21:49   0:00 grep --color=auto netbox

***@***.***:/opt/netbox/contrib$ sudo cat /etc/passwd
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
sys:x:3:3:sys:/dev:/usr/sbin/nologin
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/usr/sbin/nologin
man:x:6:12:man:/var/cache/man:/usr/sbin/nologin
lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin
mail:x:8:8:mail:/var/mail:/usr/sbin/nologin
news:x:9:9:news:/var/spool/news:/usr/sbin/nologin
uucp:x:10:10:uucp:/var/spool/uucp:/usr/sbin/nologin
proxy:x:13:13:proxy:/bin:/usr/sbin/nologin
www-data:x:33:33:www-data:/var/www:/usr/sbin/nologin
backup:x:34:34:backup:/var/backups:/usr/sbin/nologin
list:x:38:38:Mailing List Manager:/var/list:/usr/sbin/nologin
irc:x:39:39:ircd:/var/run/ircd:/usr/sbin/nologin
gnats:x:41:41:Gnats Bug-Reporting System (admin):/var/lib/gnats:/usr/sbin/nologin
nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin
systemd-network:x:100:102:systemd Network Management,,,:/run/systemd:/usr/sbin/nologin
systemd-resolve:x:101:103:systemd Resolver,,,:/run/systemd:/usr/sbin/nologin
systemd-timesync:x:102:104:systemd Time Synchronization,,,:/run/systemd:/usr/sbin/nologin
messagebus:x:103:106::/nonexistent:/usr/sbin/nologin
syslog:x:104:110::/home/syslog:/usr/sbin/nologin
_apt:x:105:65534::/nonexistent:/usr/sbin/nologin
tss:x:106:111:TPM software stack,,,:/var/lib/tpm:/bin/false
uuidd:x:107:112::/run/uuidd:/usr/sbin/nologin
tcpdump:x:108:113::/nonexistent:/usr/sbin/nologin
landscape:x:109:115::/var/lib/landscape:/usr/sbin/nologin
pollinate:x:110:1::/var/cache/pollinate:/bin/false
usbmux:x:111:46:usbmux daemon,,,:/var/lib/usbmux:/usr/sbin/nologin
sshd:x:112:65534::/run/sshd:/usr/sbin/nologin
systemd-coredump:x:999:999:systemd Core Dumper:/:/usr/sbin/nologin
dwoods3-adm:x:1000:1000:Darren Woods:/home/dwoods3-adm:/bin/bash
lxd:x:998:100::/var/snap/lxd/common/lxd:/bin/false
postgres:x:113:118:PostgreSQL administrator,,,:/var/lib/postgresql:/bin/bash
redis:x:114:119::/var/lib/redis:/usr/sbin/nologin

Thanks,
Darren Woods

On Jul 17, 2022, at 12:05 AM, Brian Candler ***@***.******@***.***>> wrote:


**THIS EMAIL IS FROM AN EXTERNAL SENDER OUTSIDE OF THE ALAMO COLLEGES.
Be cautious before clicking links or opening attachments from unknown sources. Do not provide personal or confidential information.**



How are you starting Netbox? If it's via systemd as per the current installation instructions, then the user will be set in the service file<https://smex-ctp.trendmicro.com/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fnetbox%2dcommunity%2fnetbox%2fblob%2fmaster%2fcontrib%2fnetbox.service%23L10%2dL11&umid=0ddd7268-476a-4ef9-81f0-29fabd7ce392&auth=d92b6398b351c9b13ed6363a6e9883f7e86857eb-5175e5db0dcf29a1b4b07d1a5879152c025cc2c0>.

If this Netbox was installed a long time ago - by which I mean more than 2-3 years ago - then older instructions showed running it as "www-data" user, since this already exists on many systems. And in that case you may also find Netbox being started from supervisord rather than systemd.

If your Netbox is actually working, you can always just look at running processes: the first column will give the user ID.

ps auxwww | grep netbox


You're not using docker are you? If so, obviously no user will be created on your host. Inside the container, I'm not sure - it looks like it might be using a user "unit"<https://smex-ctp.trendmicro.com/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fnetbox%2dcommunity%2fnetbox%2ddocker%2fblob%2frelease%2fdocker%2flaunch%2dnetbox.sh%23L55&umid=0ddd7268-476a-4ef9-81f0-29fabd7ce392&auth=d92b6398b351c9b13ed6363a6e9883f7e86857eb-097afac3b7205b5a6297c1f298d8ad7d45d999b5>

**Reply by candlerb** [reply]
(Aside: please can you edit your post and put three backticks (` ``` `) on a line of their own before and after the console quoted sections?  It will make it format properly)

Anyway, clearly it *isn't* being run as the netbox user:

```
***@***.***:/opt/netbox/contrib$ ps auxwww | grep netbox
www-data    1021  0.0  0.1  29632 21988 ?        S    01:17   0:07 /usr/bin/python3 /usr/local/bin/gunicorn -c /opt/netbox/gunicorn_config.py netbox.wsgi
^^^^^^^^
```

Now, systemd does have a habit of ignoring settings it can't understand, and you can check "journalctl -eu netbox" for warnings.  But it wouldn't default to "www-data", it would default to "root".

Do you see any reference to "www-data" in `/opt/netbox/gunicorn.py` ?

This is weird too:

```
root       70990  123  0.8 180428 139184 ?       R    21:49   0:02 python3 /opt/netbox/netbox/manage.py rqworker directory = /opt/netbox/netbox/ user = www-data
```

I'm afraid you'll just have to go through the full set of Netbox installation instructions, compare the settings one by one, and see what's different about your system.

Did you install it yourself, or did someone else install it?  What version of Netbox is running?  That would at least give a clue as to how old it is.

**Comment by PrieserMax** [comment]
Hello,
I'm having the same error.
I added the user to the group www-data and ran the chmod, also restarted Netbox.
Sadly no change and the same error.
Does anybody have an idea?

**Reply by candlerb** [reply]
> I'm having the same error.

You *might* be having the same error, but it's better not to make assumptions.

* Show the exact error message that you see
* State the version of Netbox you're running, and what OS you're running under

Note that Netbox instructions have not used the "www-data" user for several years, so unless your installation is ancient, the solution you've tried is almost certainly wrong.

**Reply by PrieserMax** [reply]
Thank you very much.
I'm on the latest 4.1.1
-> I'm having the error probably since installing the update to 4.1.1
![grafik](https://github.com/user-attachments/assets/8df695ec-3798-4db9-b865-3fd4a4c282d4)

In text form:
`Serverfehler

Bei der Verarbeitung dieser Anfrage wurde ein Dateiberechtigungsfehler festgestellt. Zu den häufigsten Ursachen gehören die folgenden:

Ungenügende Schreibrechte im Medienverzeichnis. Das konfigurierte Medienverzeichnis ist . Stellen Sie sicher, dass der Benutzer NetBox ausgeführt wird und Zugriff hat, um Dateien innerhalb dieses Pfads zu schreiben.

Die vollständige Ausnahme finden Sie unten.:

<class 'PermissionError'>

[Errno 13] Permission denied: '/opt/netbox/netbox/media/devicetype-images/sdgfhdsfjkf.png'

Python-Version: 3.10.12
NetBox-Version: 4.1.1
Plugins: Keine installiert

Wenn Sie weitere Unterstützung benötigen, senden Sie bitte eine E-Mail an [NetBox-Diskussionsforum](https://github.com/netbox-community/netbox/discussions) auf GitHub.
[Startseite](https://netbox.reichart.tech/)`


OS is Ubuntu server 24.04.1

I have to say, I just installed it according to the docs back then, I guess since ~version 3.7 or so and always updated it according to the instructions.

**Reply by candlerb** [reply]
> OS is Ubuntu server 24.04.1

Then it's not SELinux, and probably not apparmor.

I suggest you check:

```
ps auxwww | grep gunicorn
```

Is it running as the `netbox` user? (and you don't have multiple `netbox` users in /etc/passwd)

Then I'd check permissions as usual:

```
sudo -u netbox wc /opt/netbox/netbox/media/devicetype-images/sdgfhdsfjkf.png
ls -ld / /opt /opt/netbox /opt/netbox/netbox /opt/netbox/netbox/media /opt/netbox/netbox/media/devicetype-images /opt/netbox/netbox/media/devicetype-images/sdgfhdsfjkf.png
```

The first checks whether the `netbox` user is able to open the file, and the second shows the permissions on the file and its parent directories.

(Is the filename really "sdgfhdsfjkf.png"?)

**Reply by PrieserMax** [reply]
Thank you very much!
The output is the following:
`netbox   1540270  0.0  0.5  31876 22264 ?        Ss   Sep18   0:04 /opt/netbox-4.1.1/venv/bin/python3 /opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
netbox   1540271  0.0  5.4 289132 220196 ?       Sl   Sep18   0:10 /opt/netbox-4.1.1/venv/bin/python3 /opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
netbox   1540272  0.0  5.5 295488 220440 ?       Sl   Sep18   0:10 /opt/netbox-4.1.1/venv/bin/python3 /opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
netbox   1540273  0.0  5.4 360800 218092 ?       Sl   Sep18   0:10 /opt/netbox-4.1.1/venv/bin/python3 /opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
netbox   1540274  0.0  5.4 288112 218096 ?       Sl   Sep18   0:10 /opt/netbox-4.1.1/venv/bin/python3 /opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
netbox   1540275  0.0  5.1 278896 205444 ?       Sl   Sep18   0:10 /opt/netbox-4.1.1/venv/bin/python3 /opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
adminis+ 1601686  0.0  0.0   6612  2280 pts/0    S+   06:47   0:00 grep --color=auto gunicorn`

As far as I understand, it is running under the netbox user.
-> I only have one netbox user in /etc/passwd

Yes, that is the filename. I renamed it since I thought maybe, it is already existing on the server and can't replace it or something like that.

I looked into the folder and noticed, there are no photos available -> also checking in Netbox, there are no photos in the models. The links are still existent, but no files in that directory.
administrator@s49:/opt/netbox/netbox/media/devicetype-images$ ls
administrator@s49:/opt/netbox/netbox/media/devicetype-images$

**Reply by PrieserMax** [reply]
Update: I tried reimporting at least my old files from backup, but am not able to upload via SCP -> Permission denied

**Reply by candlerb** [reply]
> Update: I tried reimporting at least my old files from backup, but am not able to upload via SCP -> Permission denied

OK, that's a local sysadmin / permissions problem then (and probably also causes the netbox error you see).

Look in "dmesg" for errors. Check your filesystem hasn't ended up mounted read-only. Usual system admin stuff. But clearly this is not a Netbox problem now.

**Comment by PrieserMax** [comment]
Yes, probably.
Problem is, normally I'm a Windows Admin and only use it in my homelab. Will look into setting the permission correctly.

**Comment by PrieserMax** [comment]
So I chowned the directory to my user, added the old files, then chowned it to the netbox user.
Now it works


Thank you for your help!
